File: perl/pe018.pl
#!/usr/bin/perl
# http://projecteuler.net/problem=18 & http://projecteuler.net/problem=67

use strict;
use warnings;

my @triangle;

open(INPUT, '<', $ARGV[0]) or die($!);

foreach my $row (<INPUT>) {
    push(@triangle, [split(' ', $row)]);
}

close INPUT or die $!;

for (my $r = $#triangle; $r > 0; $r--) {
    for (my $c = 0; $c < $#{$triangle[$r]}; $c++) {
        my $a = $triangle[$r][$c];
        my $b = $triangle[$r][$c + 1];

        if ($a > $b) {
            $triangle[$r - 1][$c] += $a;
        }
        else {
            $triangle[$r - 1][$c] += $b;
        }
    }
}

if (!defined($triangle[0][0])) {
    exit 1;
}

print $triangle[0][0] . "\n";

File: ruby/pe018.rb
#!/usr/bin/env ruby
# http://projecteuler.net/problem=18 & http://projecteuler.net/problem=67

triangle = File.read(ARGV[0]).map { |line| line.split.map { |n| n.to_i } }
(triangle.length - 1).downto(0).each do |row|
  (0..triangle[row].length - 2).each do |column|
    cell = triangle[row][column]
    neighbor = triangle[row][column + 1]
    triangle[row - 1][column] += [cell, neighbor].max
  end
end
puts triangle[0][0]
